cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt

```diff
@@ -680,20 +681,101 @@ class ScopeManager : ScopeProvider {
      * @return a list of possible functions
      */
     @JvmOverloads
-    fun resolveFunction(
+    fun resolveFunctionLegacy(
         call: CallExpression,
         startScope: Scope? = currentScope
     ): List<FunctionDeclaration> {
         val (scope, name) = extractScope(call, startScope)
 
         val func =
             resolve<FunctionDeclaration>(scope) {
-                it.name.lastPartsMatch(name) && it.hasSignature(call)
+                it.name.lastPartsMatch(name) && it.hasSignature(call.signature, call.arguments)
             }
 
         return func
     }
 
+    class CallResolutionResult(
+        /** The original call expression. */
+        val call: CallExpression,
+        var candidateFunctions: List<FunctionDeclaration>,
```

Comment:
It's not quite clear why we need the `candidateFunctions` or the `call` in the result. It's either the input or an intermediate step, so it's not super relevant? Some doc for all the lists, sets, maps would be useful. Why is `candidateFunctions` a list and not a set (same for `bestViable`)?

Reply:
I would keep it to give the language refinement method as much information as possible, if a language decides to implement something very very accurate. Also its nice to have from a debugging perspective if you look at a finished result and see what the original candidates were.

Reply:
ok, so we keep them. Remaining points:
* Documentation of `candidateFunctions`, `viableFunctions`, `signatureResults`, `bestViable`
* `List` vs `Set`: There's no ordering and no duplicates, so a `Set` should work in all cases

Reply:
Added more documentation and made all of them a set.